StRoot/StFcsWaveformFitMaker/PeakAna.cxx

```diff
@@ -0,0 +1,1336 @@
+{
+  if( this == &rhs ){ return *this; }
+  //Note that the graph is not cloned because PeakAna should not be changing the contents of the data
+  mG_Data = rhs.GetData();//Set graph object now so doesn't get created in 'Init'
+  Init();//This sets internal signal to true so set it false after call
+  mInternalSignal=false;//Since graph is not copied don't treat as an internal signal
+  
+  ((PeakAna&)rhs).Copy(*this);
+
+  return *this;
+}
+
+//Note that this function doesn't copy the graph object or a copies the vector of peaks. This is done on purpose becuase the idea is that the copy should run its own analysis and the algorithm is robust enough to generate the same results. To copy all objects use TObject::Clone
+void PeakAna::Copy(TObject& obj) const
+{
+  TObject::Copy(obj);//Copy TObject bits
+
+  ((PeakAna&)obj).mMaxX = mMaxX;
+  ((PeakAna&)obj).mMaxY = mMaxY;
+
+  ((PeakAna&)obj).mXRangeMin = mXRangeMin;
+  ((PeakAna&)obj).mYRangeMin = mYRangeMin;
+  ((PeakAna&)obj).mXRangeMax = mXRangeMax;
+  ((PeakAna&)obj).mYRangeMax = mYRangeMax;
+  ((PeakAna&)obj).mBaseline = mBaseline;
+  ((PeakAna&)obj).mBaselineSigma = mBaselineSigma;
+  ((PeakAna&)obj).mBaselineSigmaScale = mBaselineSigmaScale;
+  ((PeakAna&)obj).mSearch.SetWindow(SearchPeak(),SearchWidth());
+  ((PeakAna&)obj).mDeltaX = mDeltaX;
+  ((PeakAna&)obj).mFilter = mFilter;
+  ((PeakAna&)obj).mFilterScale = mFilterScale;
+  //((PeakAna&)obj).mFilterSigma = mFilterSigma;
+  if( ((PeakAna&)obj).mFilterWeights!=0 ){ delete [] ((PeakAna&)obj).mFilterWeights; ((PeakAna&)obj).mFilterWeights = 0; }
+  if( mFilterWeights!=0 && mFilterScale>0 ){
+    ((PeakAna&)obj).mFilterWeights = new double[2*mFilterScale+1];
+    std::copy( mFilterWeights,mFilterWeights+(2*mFilterScale+1), ((PeakAna&)obj).mFilterWeights );
+  }
+  else{ ((PeakAna&)obj).mFilterWeights = 0; }
+  ((PeakAna&)obj).mTunnelScale = mTunnelScale;
+  ((PeakAna&)obj).mTunnelSigma = mTunnelSigma;
+  ((PeakAna&)obj).mTunnelThreshold = mTunnelThreshold;
+  
+  ((PeakAna&)obj).mPainter = 0;//Dont' copy painter
+  
+  return;  
+}
+
+TObject* PeakAna::Clone(const char* newname) const
+{
+  TGraph* cloneg = (TGraph*)this->GetData()->Clone(newname);
+  PeakAna* ana = new PeakAna(cloneg);
+  Copy(*ana);
+  ana->ForceInternal();//Cloned graphs should always be internal
+  if( this->FoundPeakIndex()>=0 ){
+    //If mComputedIndex>=0 this means AnalyzeForPeak() was called and need to copy the peaks. This is preffered over re-running the analysis as copy should be faster than re-running the data
+    ana->mComputedIndex = this->mComputedIndex;
+    for( UInt_t i=0; i<mPeaks.size(); ++i ){
+      ana->mPeaks.push_back( mPeaks.at(i) );
+    }
+    ana->mFoundPeak = this->mFoundPeak;
+  }
+  return ana;
+}
+
+PeakAna::~PeakAna()
+{
+  if(mG_Data!=0 && mInternalSignal ){delete mG_Data;}
+  if( mFilterWeights!=0 ){ delete [] mFilterWeights; mFilterWeights=0; }
+  if( mPainter!=0 ){ delete mPainter; mPainter=0; }
```

Comment:
The checks for nullptr are unnecessary

```suggestion
  if (mInternalSignal) { delete mG_Data; mG_Data= nullptr; }
  delete [] mFilterWeights; mFilterWeights = nullptr;
  delete mPainter; mPainter = nullptr;
```

Reply:
I think they are necessary because, even though I tried to always keep a graph object alive, it is a sanity check in case someone decides to call *SetData(0)* in which case mG_Data will equal 0 and then you are deleting a nullptr which is bad. The same goes for the others since mFilterWeights=0 unless *SetFilter(2,...)* is called, and mPainter=0 unless *Paint()* is called. I don't remember why I don't check for a nullptr in *SetData()* so I'll add that and then I agree that `mG_Data!=0` is unnecessary. Also, since this is in the destructor I probably don't need to set them to the nullptr afterwords.

Reply:
It's fine to do `x=nullptr; delete x;`
See e.g. https://en.cppreference.com/w/cpp/language/delete

> If expression evaluates to a null pointer value, no destructors are called, and the deallocation function may or may not be called (it's unspecified), but the default deallocation functions are guaranteed to do nothing when passed a null pointer.

Reply:
@veprbl I didn't know that. Thanks! this is great to know. I'll change it.